src/main/java/com/microsoft/sqlserver/jdbc/SQLQueryMetadataCache.java

```diff
@@ -0,0 +1,272 @@
+                                                    ("R_cryptoCacheInaccessible"), null, 0, false);
+            }
+        }
+        
+        // If the size of the cache exceeds the threshold, set that we are in trimming and trim the cache accordingly.
+        int cacheSizeCurrent = session.getCryptoCache().getParamMap().size();
+        if (cacheSizeCurrent > cacheSize + cacheTrimThreshold) {
+            try {
+                int entriesToRemove = cacheSizeCurrent - cacheSize;
+                ConcurrentHashMap<String, ConcurrentHashMap<String, CryptoMetadata>> newMap = new ConcurrentHashMap<>();
+                ConcurrentHashMap<String, ConcurrentHashMap<String, CryptoMetadata>> oldMap = 
+                    session.getCryptoCache().getParamMap();
+                int count = 0;
+
+                for (Map.Entry<String, ConcurrentHashMap<String, CryptoMetadata>> entry : oldMap.entrySet()) {
+                    if (count >= entriesToRemove) {
+                        newMap.put(entry.getKey(), entry.getValue());
+                    }
+                    count++;
+                }
+                session.getCryptoCache().replaceParamMap(newMap);
+
+            } catch (Exception e) {
+                throw new SQLServerException(null, SQLServerException.getErrString
+                                    ("R_cryptoCacheInaccessible"), null, 0, false);
+            }
+        }
+        
+        session.getCryptoCache().addParamEntry(encryptionValues.getKey(), metadataMap);
+        return true;
+    }
+
+    
+    /**
+    * 
+    * Remove the cache entry.
+    * 
+    * @param stmt
+    *        SQLServer statement used to retrieve keys
+    * @param session
+    *        The enclave session where the cryptocache is stored
+    * @param connection
+    *        The SQLServerConnection, also used to retrieve keys
+    */
+    static void removeCacheEntry(SQLServerStatement stmt, EnclaveSession session,
+            SQLServerConnection connection) {
+        AbstractMap.SimpleEntry<String, String> encryptionValues = getCacheLookupKeys(stmt, connection);
+        if (encryptionValues.getKey() == null) {
+            return;
+        }
+
+        session.getCryptoCache().removeParamEntry(encryptionValues.getKey());
+    }
+
+    
+    /**
+    * 
+    * Returns the cache and enclave lookup keys for a given connection and statement
+    * 
+    * @param statement
+    *        The SQLServer statement used to construct part of the keys
+    * @param connection
+    *        The connection from which database name is retrieved
+    * @return
+    *        A key value pair containing cache lookup key and enclave lookup key
+    */
+    private static AbstractMap.SimpleEntry<String, String> getCacheLookupKeys(
+            SQLServerStatement statement, SQLServerConnection connection) {
+        final int sqlIdentifierLength = 128;
+        
+        if (connection == null) {

```

Comment:
when would connection be null? I would think something bad has happened wondering if we should fail or at least log a warning?

Reply:
Correct, there's no case it would be null. If it was null we wouldn't even be able to call the function, this has been removed. I initially added it as I planned to use addQueryMetadata in testing, but decided against it.